dkg-gadget/src/keygen_manager/mod.rs

```diff
@@ -0,0 +1,385 @@
+}
+
+// only 1 task at a time may run for keygen
+const MAX_RUNNING_TASKS: usize = 1;
+
+impl<B, BE, C, GE> KeygenManager<B, BE, C, GE>
+where
+	B: Block,
+	BE: Backend<B> + Unpin + 'static,
+	GE: GossipEngineIface + 'static,
+	C: Client<B, BE> + 'static,
+	C::Api: DKGApi<B, AuthorityId, NumberFor<B>, MaxProposalLength, MaxAuthorities>,
+{
+	pub fn new(logger: DebugLogger, clock: impl HasLatestHeader<B>) -> Self {
+		Self {
+			work_manager: WorkManager::<B>::new(
+				logger,
+				clock,
+				MAX_RUNNING_TASKS,
+				PollMethod::Manual,
+			),
+			active_keygen_retry_id: Arc::new(AtomicUsize::new(0)),
+			keygen_state: Arc::new(Atomic::new(KeygenState::Uninitialized)),
+			latest_executed_session_id: Arc::new(Atomic::new(None)),
+			finished_count: Arc::new(AtomicUsize::new(0)),
+			_pd: Default::default(),
+		}
+	}
+
+	pub fn deliver_message(&self, message: SignedDKGMessage<Public>) {
+		let message_task_hash =
+			*message.msg.payload.keygen_protocol_hash().expect("Bad message type");
+		self.work_manager.deliver_message(message, message_task_hash)
+	}
+
+	pub fn session_id_of_active_keygen(&self, now: NumberFor<B>) -> Option<JobMetadata> {
+		self.work_manager.get_active_session_ids(now).pop()
+	}
+
+	pub fn get_latest_executed_session_id(&self) -> Option<SessionId> {
+		self.latest_executed_session_id.load(Ordering::SeqCst)
+	}
+
+	fn state(&self) -> KeygenState {
+		self.keygen_state.load(Ordering::SeqCst)
+	}
+
+	pub fn set_state(&self, state: KeygenState) {
+		self.keygen_state.store(state, Ordering::SeqCst);
+	}
+
+	#[allow(clippy::needless_return)]
+	pub async fn on_block_finalized(
+		&self,
+		header: &B::Header,
+		dkg_worker: &DKGWorker<B, BE, C, GE>,
+	) {
+		if let Some((active, _queued)) = dkg_worker.validator_set(header).await {
+			// poll to clear any tasks that have finished and to make room for a new potential
+			// keygen
+			self.work_manager.poll();
+			let now_n = *header.number();
+			let block_id: u64 = now_n.saturated_into();
+			let session_id = active.id;
+			let current_protocol = self.session_id_of_active_keygen(now_n);
+			let state = self.state();
+			let should_execute_new_keygen = dkg_worker.should_execute_new_keygen(header).await;
+			let executed_count = self.finished_count.load(Ordering::SeqCst);
+			dkg_worker.logger.debug(format!(
+				"*** KeygenManager on_block_finalized: session={session_id},block={block_id}, state={state:?}, current_protocol={current_protocol:?} | total executed: {executed_count}",
+			));
+			dkg_worker
+				.logger
+				.debug(format!("*** Should execute new keygen? {should_execute_new_keygen:?}"));
+
+			/*
+			   Session 0 (beginning): immediately run genesis
+			   Session 0 (ending): run keygen for session 1
+			   Session 1 (ending): run keygen for session 2
+			   Session 2 (ending): run keygen for session 3
+			   Session 3 (ending): run keygen for session 4
+			*/
+
+			/*
+			   Pre checks
+			*/
+
+			// it's possible genesis failed and we need to retry
```

Comment:
Please capitalize doc strings if they're sentences.
```suggestion
			// It's possible genesis failed, and we need to retry
```